File: spark_basic/task1.py
# Wentao Zhou
# DSCI 553 - hw1, task1, spring 2021
# spark warm-up practice

from pyspark import SparkContext
import json
import sys

sc = SparkContext('local[*]', 'task1')
input_file_path = sys.argv[1]
output = sys.argv[2]
textRDD = sc.textFile(input_file_path).map(json.loads)\
    .map(lambda row: (row["business_id"], row["user_id"], row["date"])).cache()
totalnumReview = textRDD.count() #q1
numreview = textRDD.filter(lambda row: int(row[2][0: 4]) == 2018).count() #q2
#textRDD = textRDD.filter(lambda dic: len(dic["text"]) > 0)
numofdistinctuser = textRDD\
    .map(lambda row: row[1]).distinct().count()  #q3
toptenusers = textRDD\
    .map(lambda row: [row[1], 1]).reduceByKey(lambda a, b: a+b).takeOrdered(10, key=lambda x: (-x[1], x[0])) #Q4
numdistinctbusineess = textRDD\
    .map(lambda row: row[0]).distinct().count() #q5
toptenbusiness = textRDD\
    .map(lambda row: [row[0], 1]).reduceByKey(lambda a, b: a+b).takeOrdered(10, key=lambda x: (-x[1], x[0])) #q6

res_dic = {"n_review": totalnumReview,
            "n_review_2018": numreview,
            "n_user": numofdistinctuser,
            "top10_user": toptenusers,
            "n_business": numdistinctbusineess,
            "top10_business": toptenbusiness}
with open(output, 'w') as output:
    json.dump(res_dic, output)
output.close()

File: spark_basic/task1.scala
import org.apache.spark.{SparkConf, SparkContext}
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.json4s.JsonDSL._
/*
object task1 {

  def main(arg: Array[String]): Unit={
    val spark = new SparkConf().setAppName("task1").setMaster("local[*]")
    val sc = new SparkContext(spark)
    val input_file = "/Users/wentaozhou/desktop/test_review.json"
    val rdd = sc.textFile(path = input_file)
    val textRDD = rdd.map{ row => val json_row = jsonStrToMap(row)
      (json_row)
    }
    val newRDD = textRDD.map(row => (row("user_id"), row("business_id"), row("date")))
    val count = newRDD.count()
   // val review2018 =  textRDD.filter(row => row("date").toString().contains("2018")).count()
    val review = newRDD.filter(row => row._3.toString.contains("2018")).count()
    val distinctuser = newRDD.map(row => row._1.toString()).distinct().count()
    val toptenusers = newRDD.map(row => (row._1,1)).reduceByKey((a,b) => a+b)
      .sortBy(row => (-row._2.toInt, row._1.toString())).take(10)
    val numdistinctbusineess = newRDD.map(row => row._2).distinct().count()
    val toptenbusiness = newRDD.map(row => (row._2,1)).reduceByKey((a,b) => a+b).sortBy(row => (-row._2.toInt, row._1.toString())).take(10)
    val json = ("n_review" -> count) ~ ("n_review_2018" -> toptenusers)
    print(compact(render(json)))
  }



  def jsonStrToMap(jsonStr: String): Map[String, Any] = {
    implicit val formats = org.json4s.DefaultFormats
    parse(jsonStr).extract[Map[String, Any]]
  }
}
*/